### PR TITLE
Avoid unwanted path stripping

### PR DESCRIPTION
**Pull request by OskarD**
#3764

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@OskarD Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=3765)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=3765) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@OskarD Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=3765)!

**Comment by OskarD** [comment]
I'm guessing the test will fail with something like this:

> Expecting:
 <Route{id='foo-bar', fullPath='/foo-bar/1', path='/1', location='foo', prefix='/foo-bar', retryable=false, sensitiveHeaders=[], customSensitiveHeaders=false, prefixStripped=true}>
to be equal to:
 <Route{id='foobar', fullPath='/foo-bar/1', path='/1', location='foo', prefix='/foo-bar', retryable=false, sensitiveHeaders=[], customSensitiveHeaders=false, prefixStripped=true}>
but was not.

But I don't understand why? Why does the ID suddenly have the hyphen?

**Comment by OskarD** [comment]
> I'm guessing the test will fail with something like this:
> 
> > Expecting:
> > <Route{id='foo-bar', fullPath='/foo-bar/1', path='/1', location='foo', prefix='/foo-bar', retryable=false, sensitiveHeaders=[], customSensitiveHeaders=false, prefixStripped=true}>
> > to be equal to:
> > <Route{id='foobar', fullPath='/foo-bar/1', path='/1', location='foo', prefix='/foo-bar', retryable=false, sensitiveHeaders=[], customSensitiveHeaders=false, prefixStripped=true}>
> > but was not.
> 
> But I don't understand why? Why does the ID suddenly have the hyphen?

It came true!!! But why?

**Comment by spencergibb** [comment]
Looking good. Is there an edge case where just `/foo` (ie context) is routed thru zuul? Maybe test that and see what happens.

**Comment by OskarD** [comment]
> Looking good. Is there an edge case where just `/foo` (ie context) is routed thru zuul? Maybe test that and see what happens.

I don't understand why `SimpleRouteLocator::getRoute` is mocked, but it seems to be working.

I looked closer at the ID thing, and I get it now... although it is a little confusing

**Comment by OskarD** [comment]
> Looking good. Is there an edge case where just `/foo` (ie context) is routed thru zuul? Maybe test that and see what happens.

I don't understand why `SimpleRouteLocator::getRoute` is mocked, but it seems to be working.

I looked closer at the ID thing, and I get it now... although it is a little confusing

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java [hunk 1]
```diff
@@ -216,7 +216,7 @@ private String adjustPath(final String path) {
 		if (RequestUtils.isDispatcherServletRequest()
 				&& StringUtils.hasText(this.dispatcherServletPath)) {
 			if (!this.dispatcherServletPath.equals("/")
-					&& path.startsWith(this.dispatcherServletPath)) {
+					&& path.startsWith(this.dispatcherServletPath + "/")) {
```
Comment (spencergibb) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java:219]: This never gets exercised by your new tests since `RequestUtils.isDispatcherServletRequest()` is always false.

Comment (spencergibb) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java:219]: Adding
```
		RequestContext.getCurrentContext().set(IS_DISPATCHER_SERVLET_REQUEST_KEY, true);
// test
		RequestContext.getCurrentContext().remove(IS_DISPATCHER_SERVLET_REQUEST_KEY);
```
reveals that `test_getMatchingRouteFilterWithEqualServletPath()` fails with a null route

Comment (OskarD) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java:219]: What should be the expected behavior?
As I understand it, the purpose of this method is to remove the `dispatcherServletPath` from the path if someone puts it there. Isn't it better to just let it stay wrong, so the user can fix their bad configuration?

Comment (spencergibb) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java:219]: a dispatcher path of `/bar` and a routepath of `/bar/**` will only match if the request path is `/bar/bar`. The tests need to be updated to add the request context setup I mentioned above and either add another test that matches a default route (ie dispatcher path of `/bar` and routepath of `/**`) or change the test that returns null to the values I described.

Comment (OskarD) [reply spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/SimpleRouteLocator.java:219]: I have changed the test to expect null

**Comment by OskarD** [comment]
Why is codecov/project failing? When I click details I get "Unable to find report content in the storage archive."

**Comment by spencergibb** [comment]
Don't worry about codecov

**Comment by OskarD** [comment]
@spencergibb will this be merged?

**Comment by OskarD** [comment]
@spencergibb ping

**Comment by spencergibb** [comment]
I haven't had a chance to review it since the updates.

**Comment by spencergibb** [comment]
The new tests still don't fail if I remove the change to `SimpleRouteLocator`. Without this I can't tell if there is a regression or now.
